### ISSUE TITLE
Request body can't be easily restricted to a trait

### ISSUE THREAD

**Issue by sagebind** [hyperium/http#88]
I'm aware that the title of this issue reads really weird, so I will do my best to explain and start a discussion. My explanation is a little long, so if you're in a hurry here's the TL;DR:

As a server application writer, there's no way I can handle a `Request<T>` without knowing what `T` is. As a gateway/framework, there's no way I can enforce what `T` is without making the application non-interoperable with other gateways/frameworks.

---

I've been working on a web application interface crate (think Ruby's Rack) for a couple of months, but I heard that this crate was in the works and decided to wait and see what this crate would look like. (I haven't publicly announced said crate, but I guess the cat is out of the bag now...)

I am inspecting it now to see how well this crate will work as an interface for HTTP applications, and I am encountering a problem with `Request`. The problem is that the request is generic over `T`, which is the body. There is no way for an application to realistically read a body of unknown type.

Consider the following web application (greatly simplified, but works as an example):

```rust
use http::{Request, Response};

pub fn my_application<Body>(request: Request<Body>) -> Response<&'static str> {
    Response::new("Hello World!".into())
}
```

That works great, but what if we want to do something more interesting, like respond with the number of bytes in the request body? That should be easy:

```rust
use http::{Request, Response};
use std::io::Read;

pub fn my_application<Body: Read>(mut request: Request<Body>) -> Response<String> {
    let mut buf = Vec::new();
    let len = request.body_mut().read_to_end(&mut buf).unwrap();

    Response::new(format!("{}", len))
}
```

That should work, but now we have a problem: we can no longer handle any type of request; we can only handle requests of a specific type. More specifically, we accept a request of type `http::Request<T> where T: std::io::Read`. If we wanted to support `Request<String>`, we'd have to re-implement our application function again for this type.

It gets worse. The above examples are stateless, which very few real-world applications are. We'd probably want to instead talk about these applications as traits instead of functions. Something like this comes to mind (more similar to my crate, actually similar to Iron too):

```rust
use http::{Request, Response};

pub trait HttpApp: Send + Sync + 'static {
    fn handle<Req, Res>(&self, _: Request<Req>) -> Response<Res>;
}
```

There's an immediate problem with this; no HTTP server implementation can actually support the above trait. If I write a CGI gateway crate, I might provide a `Request<Stdin>`, but now the application writers have to explicitly code against `Request<Stdin>`, and interoperability is lost.

To put it another way: `Request<T>` doesn't provide application writers a unified request type to code against, because `Request` is generic over something without meaningful bounds. If application writers do code to a specific body type, then they are no longer coding against a crate-agnostic type. `T` here is basically leaking the implementation into the interface.

I don't pretend to have an answer to this problem; maybe this crate is too high-level (or is it low-level?) for what I need. I am very interested in comments, disagreements, or suggestions.

**Comment by seanmonstar** [comment]
Thanks for the detailed write up! We've also been interested in something like WSGI/Rack/Servlet. Hopefully with the community's help, we can make something stellar.

----

In regards to `T` having no bounds: **This is actually a benefit!**

First of all, by not having bounds on the body generic, this allows the crate to be used in both asynchronous frameworks and in synchronous. Requiring a trait bounds would likely hamper one of those sides. Likely, framework ecosystems will show up that do require a bounds. You could imagine a framework that behaves somewhat like hyper today, where `T: Stream` is required.

Secondly, this allows for application writers to have these bounds at the borders, but to then map them into anything they want internally. So, if an framework receives `Request<T> where T: Stream<Item=Vec<u8>, Error=io::Error>`, a user could map the request into something like a `Request<Comment>` internally. Something like this:

```rust
fn framework_entrypoint<T>(req: Request<T>) -> BoxFuture<Response<T>, io::Error>
where T: Stream<Item=Vec<u8>, Error=io::Error> {
    if req.method() == POST && req.uri().path() == "/article/comment" {
        let (head, body) = req.into_parts();
        body.concat2().map(move |vec| {
            let comment = serde_json::from_slice(&vec)?;
            let req = Request::from_parts(head, comment);
            post_comment_for_article(req)
        })
    }
    // ...
}

#[derive(Deserialize)]
struct Comment {
    author: String,
    body: String,
}

fn post_comment_for_article(req: Request<Comment>) -> Something {
    // ...
}
```

Additionally, any sort of middleware that only adjusts headers can be made to work for any framework, since it can just put zero bounds on the body (since it won't look at the body anyways).

```rust
fn apply_cors<T>(res: &mut Response<T>, options: &CorsOptions) {
    res.headers_mut().insert(name, value);
}
```

**Comment by carllerche** [comment]
It looks like the discussion here came to a close.

**Comment by jimmycuadra** [comment]
We're discussing this issue right now in the Ruma project. We use an [Endpoint](https://docs.rs/ruma-api/0.5.0/ruma_api/trait.Endpoint.html) trait for each API endpoint which converts between `http` request and response types and our own concrete request and response types defined for each API endpoint. We have a crate (ruma-client-api) that implements these conversions, and a separate crate (ruma-client) that actually provides an HTTP client using Hyper. Notably, Ruma is only bound to Hyper at the level of ruma-client.

Our goal is to keep ruma-api and ruma-client-api agnostic to HTTP client implementations. Without the ability to constrain what type "body" is, and in the absence of any convention in the Rust HTTP ecosystem, there's no way to:

1. Implement our `Endpoint` trait, because we need _some_ knowledge of what the body is 2. Keep it interoperable with potential other HTTP clients.

To make item 1 possible, we'd need to add a restriction that the body is a `Stream` whose items can be deref coerced into `[u8]`. That would make it work with Hyper's `Body` and `Chunk` types. But we have no way of knowing whether or not such an interface would be compatible with other HTTP clients. Perhaps a convention will emerge in the ecosystem, but the unconstrained generic type in `http` means that there's no real way to future proof this.

**Comment by carllerche** [comment]
The HTTP crate specifically does not want to constrain request / response bodies to anything that resembles bytes. This would prevent storing a typed representation of the body.

**Comment by jimmycuadra** [comment]
Yep, I understand the reason it's designed the way it is. But the issue brought up here is a cost of this design. The crate offers types for everything except req/res bodies, and as a result req/res bodies cannot be interoperable between higher level crates that both use HTTP.
